File: src/parser/expressions/additive-expression.cpp
#include "parser/expressions/additive-expression.h"
#include "parser/expressions/multiplicative-expression.h"
#include "parser/node.h"

static Node* make(Node& left, Token& op, Node& right) {
    return new AdditiveExpression(left, op, right);
}


AdditiveExpression::AdditiveExpression(Node& left, Token& op, Node& right)
: BinaryOperation(left, op, right) {}

Node* AdditiveExpression::construct(Parser& parser) {
    return BinaryOperation::construct(
        parser,
        {"+", "-"}, 
        MultiplicativeExpression::construct, 
        make);
}

File: app/src/main/kotlin/sea/parser/expressions/AdditiveExpression.kt
package sea.grammar

import sea.parser.*
import sea.grammar.MultiplicativeExpression

class AdditiveExpression(left: Node, op: Token, right: Node)
    : BinaryOperation(left, op, right) {
        companion object : Node.CompanionObject {
            override fun construct(parser: Parser): Node {
                @Suppress("UNCHECKED_CAST")
                val cls = AdditiveExpression::class as BinOpClass
                val hasList = listOf("+", "-")

                return construct(parser, hasList, MultiplicativeExpression::construct, cls)
            }
        }

        override fun transpile(transpiler: Transpiler): TExpression {
            return transpiler.nodeContext(this) {
                val left = left.transpile(transpiler).arithmeticOp(transpiler)
                val right = right.transpile(transpiler).arithmeticOp(transpiler)
                val result = TExpression.resolveType(left, right)

                if(left.longValue != null && right.longValue != null) {
                    if(op.has("+")) result.longValue = left.longValue!! + right.longValue!!
                    else result.longValue = left.longValue!! - right.longValue!!
                }

                result.replace("$left ${op.string} $right")
            }
         }
    }
